README.md
SNMP
----

[![Chef cookbook](https://img.shields.io/cookbook/v/chef-sugar.svg)](https://github.com/thomasvincent/cookbook-snmp)
[![Build Status](https://travis-ci.org/thomasvincent/cookbook-snmp.svg?branch=master)](https://travis-ci.org/thomasvincent/cookbook-snmp)

Installs and configures snmpd.

### SNMP multiplex

The SNMP multiplex (smuxpeer) line will be set for Dell OpenManage, if Dell
manufactured hardware is detected by Ohai.

### Recipes

* snmp::default
  - Installs and configures SNMPD

* snmp::snmptrapd
  - Configures snmptrapd

### Attributes - snmp - global configuration

* `snmp['snmp']['mibdirs']`
  - MIB directories.  
  - Default: "/usr/share/snmp/mibs".

* `snmp['snmp']['mibs']`
  - MIBs to load.  The default is not to set any and is therefore dependant
     on the daemon default.
  - Default: nil.

### Attributes - snmpd - basis

Notable overridable attributes are as follows.  It is recommended to override
these following attributes to best suit your own environment.

* `snmp['snmpd']['conffile']`
  - ...
  - Default: 

* `snmp['snmpd']['conf_owner']`
  - ...
  - Default: 

* `snmp['snmpd']['conf_group']`
  - ...
  - Default: 

* `snmp['snmpd']['conf_mode']`
  - ...
  - Default: 

### Attributes - snmpd - debian specific configuration

* `snmp['snmpd']['snmpd_run']`
  - snmpd control. "yes" means start daemon)
  - Default: "yes".

* `snmp['snmpd']['snmpd_opts']`
  - snmpd options
  - Default: 

* `snmp['snmpd']['trapd_run']`
  - snmptrapd control (default of "no" means do not start daemon)
    master agentx support must be enabled in snmpd before snmptrapd
    can be run.  See snmpd.conf(5) for how to do this.
  - Default: 

* `snmp['snmpd']['trapd_opts']`
  - snmptrapd options
  - Default: 

* `snmp['snmpd']['snmpd_compat']`
  - Create symlink on Debian legacy location to official RFC path
  - Default: is "yes".

### Attributes - inside snmpd config

* `snmp['snmpd']['agentAddress']`
  - Listen address from snmpd for ipv4
  - Default: "udp:161".

* `snmp['snmpd']['agentAddress6']`
  - Listen address from snmpd for ipv6
  - Default: "udp:161".

* `snmp['snmpd']['community']`
  - SNMP Community String
  - Default: "public".

* `snmp['snmpd']['sec_name']`
  - Hash of security name and IPv4 networks to allow access.  See `man 5 snmpd.conf`,
    section VACM Configuration for further details.
  - Default: { 'notConfigUser' => %w(default) }

* `snmp['snmpd']['sec_name6']`
  - Hash of security name and IPv6 networks to allow access.  See `man 5 snmpd.conf`,
    section VACM Configuration for further details.
  - Default: { 'notConfigUser' => %w(default) }

* `snmp['snmpd']['groups']['v1']`
  - Hash of security groups to map security names into for SNMPv1.  See `man 5 snmpd.conf`,
    section VACM Configuration for further details.
  - Default: { 'notConfigGroup' => %w(notConfigUser) }

* `snmp['snmpd']['groups']['v2c']`
  - Hash of security groups to map security names into for SNMPv2c.  See `man 5 snmpd.conf`,
    section VACM Configuration for further details.
  - Default: { 'notConfigGroup' => %w(notConfigUser) }

* `snmp['snmpd']['full_systemview']`
  - Boolean to include the full systemview.
    This defaults to "false" as many distributions ship this way to speed up
     snmpwalk.  However, if you're running SNMP Network Management System,
     you'll want to override this as "true" on your systems.
  - Default: false.

* `snmp['snmpd']['additional_oids']`
  - String or Array of Strings representing the oid to include in systemview.
    This defaults to "nil". You can add one oid in the of '.1.3.6.1.2.1.2021' or
    an array of oids like ['.1.3.6.1.2.1.2021','.1.3.6.1.4.1.20267.200']
  - Default: nil

* `snmp['snmpd']['syslocationVirtual']`
  - String to set the location for Virtual Machines.
  - Default: "Virtual Server".

* `snmp['snmpd']['syslocationPhysical']`
  - String to set the location for physical systems.
  - Default: "Server Room".

* `snmp['snmpd']['syscontact']`
  - String to set a name, and e-mail address for systems.
  - Default is "Root <root@localhost>"

* `snmp['snmpd']['sysname']`
  - Systems name. For example servername.
  - Default: automatically gathered. Nodes fqdn name. But you can override it if you want.

* `snmp['snmpd']['pass']`
  - Pass-through scripts as { oid => script } pairs.
  - Default: is an empty hash.

* `snmp['snmpd']['pass_persist']`
  - Persistent pass-through scripts as { oid => script } pairs.
  - Default: is an empty hash.

* `snmp['snmpd']['process_monitoring']['proc']`
  - Array of processes to monitor. 
  - Default: This is an empty array by default.

* `snmp['snmpd']['process_monitoring']['procfix']`
  - Array of procfix lines to configure. These are run when a monitored process is in an error state.
  - Default: This is an empty array by default.

* `snmp['snmpd']['include_all_disks']`
  - Boolean to include disk usage checks for all filesystems.
  - Default: false.

* `snmp['snmpd']['all_disk_min']`
  - Number of bytes that should be free on each disk when :include\_all\_disks 
    is 'true.'  Can also be expressed as a string specifying a minimum percent
    e.g. "10%".
  - Default: is 100 (100kB) which is the normal snmpd default.

* `snmp['snmpd']['disks']`
  - Specify individual disks to monitor or, if monitoring all disks, specify 
    different minimums.  If set, the template expects this to be an 
    array of hashes specifying the mount point of the disk to monitor and
    the minimum threshold in the form of {:mount => '/', :min => 100000}.
    As with :all\_disk\_min, :min can also be expressed as a percent.
  - Default: is an empty array.

* `snmp['snmpd']['ignoredisks']`
  -  ...
  - Default: is an empty array.

* `snmp['snmpd']['skipNFSInHostResources']`
  -  ...
  - Default: nil.

* `snmp['snmpd']['load_average']`
  - monitors the load average of the local system, specifying thresholds for the
    1-minute, 5-minute and 15-minute averages.
    If set, the template expects this to be a hash specifying the minimum
    threshold in the form of { :max1 => '12', :max5 => '14', :max15 => '14'}.
    Default is an empty array.

* `snmp['snmpd']['trapsinks']`
  - Array of trapsink hosts, and optionall Community Trap strings.
    This is an empty array by default.
  - Default: 

* `snmp['snmptrapd']['trapcommunity']`
  - SNMP Community Trap String, default is "public".
  - Default: "public".

* `snmp['snmptrapd']['trapd_run']`
  - snmptrapd control (default of "no" means do not start daemon)
    master agentx support must be enabled in snmpd before snmptrapd
    can be run.  See snmpd.conf(5) for how to do this.
  - Default: "no".

* `snmp['extend_scripts']`
  - A hash that specifies extension scripts to the SNMP daemon. If set, the template
    will place these name => value pairs in the "Extend" section of the snmpd.conf file.
    The array must be in the format of { 'foo' => '/path/to/foo -arg', 'bar' => '/path/to/bar' }
  - Default:

* `snmp['disman_events']['enable']`
  - Boolean value for enabling DisMan Event MIB configuration
  - Default is false.

* `snmp['disman_events']['user']` and `snmp['disman_events']['password']`
  - String values for SNMPv3 credentials to set for the DisMan Event MIB
    to work.
  - Default:

* `snmp['disman_events']['linkUpDownNotifications']`
  - String value which must be "yes" or "no" to enable traps for link
    state change notifications.
  - Default: is "yes".

* `snmp['disman_events']['defaultMonitors']`
  - String value which must be "yes" or "no" to enable traps for the
    DisMan Event MIB default checks. See the snmpd.conf manual for details.
  - Default: is "yes".

* `snmp['disman_events']['monitors']`
  - Array of string values for additional 'monitor' config entries.
  - Default: is an empty array.

### Usage

Here is a full example featuring commonly overrided attributes.
You can apply these override attributes in a role, or node context.

```
  override_attributes "snmp" => {
    "community" => "secret",
    "full_systemview" => true,
    "disks" => [{:mount => "/", :min => "10%"}],
    "load_average" => {:max1 => "24", :max5 => "12", :max15 => "6" },
    "trapsinks" => [ "zenoss.example.com", "nagios.example.com" ],
    "syslocationPhysical" => "Server Room",
    "syslocationVirtual" => "Cloud - Virtual Pool",
    "syscontact" => "sysadmin@example.com"
  }
```

```
{
  "name": "module_snmp",
  "description": "Baseline role applied to all servers in local network",
  "override_attributes": {
    "snmp": {
      "snmpd": {
        "community": "MYSECRET",
        "full_systemview": true,
        "group": "v2c",
        "snmpd_opts": "-LS0-5d -Lf /dev/null -p /var/run/snmpd.pid",
        "sec_name": {
            "notConfigUser": [ ],
            "SECNAME_SOME": [
              "192.168.99.88"
            ]
        },
        "sec_name6": {
          "notConfigUser": [ ]
        },
        "groups": {
          "v1": {
            "notConfigGroup" : [ ]
          },
          "v2c": {
            "notConfigGroup" : [ ],
            "GROUP_SOME": [
              "SECNAME_SOME"
            ]
          }
        },
        "ignoredisk": [
          "/media/trash"
        ],
        "skipNFSInHostResources": true,
        "include_all_disks": true,
        "all_disk_min": "10%"
      }
    }
  },
  "run_list": [
    "recipe[snmp]"
  ]
}
```


### Author and License

Author:: Eric G. Wolfe (eric.wolfe@gmail.com) [![endorse](https://api.coderwall.com/atomic-penguin/endorsecount.png)](https://coderwall.com/atomic-penguin)
Author:: Thomas Vincent (thomasvincent@gmail.com) [![endorse](https://api.coderwall.com/thomasvincent/endorsecount.png)](https://coderwall.com/thomasvincent)
Author:: Florian Asche (git@florian-asche.de)


Copyright Thomas Vincent 2017
Copyright Eric G. Wolfe 2010-2017

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
snmp cookbook CHANGELOG
=======================
v5.0.2
------
* allow a config to skip connection logging (dontLogTCPWrappersConnects)

v5.0.0
------
@florian-asche
* Cleanup configuration variables
* Fixed several bugs
* Added default configuration file template for rhel and auto restart for changes of those
* Added some missing variables
* Improved documentation

v4.0.2
------
@floored1585
* Added pass and pass_persist attributes

v4.0.1
------
* Ownership transfered to @thomasvincent
* Copyright updated
* .github templates
* configuration files updated and added
* Oracle added as supported platform

v4.0.0
------
  *BREAKING CHANGE*
  * Allow multiple sources and groups
    - Deprecate sources & sources6 keys
    - Added sec_name, sec_name6, groups['v1'], groups['v2c'] keys.

  * Added sysName to System contact information section
  * snmpd.conf should only be readable to root
  * Update various build files

v3.0.1
------
  * Missing newline in template, issue #28 reported by @indigo423

v3.0.0
------
  @atomic-penguin
  * Update build files, and rubocopped.

  @jhmartin
  * removed fqdn fron template.

  @CBeerta
  * Added `sources` and `sources6` attributes.
  * Added ipv6 listener.

Breaking changes:

  @glensc
  * Change `load_average` from Array of Hashes, to Hash #27

v2.0.0
------
  @MrMMorris
  * Add load average checks.

Breaking changes:

  @themoore
  * Remove hardcoded extend scripts and is_dnsserver boolean.
  * Remove perl dependency.
  * Added extend_scripts hash for greater flexibility. #21

v1.1.0
------
  @slashterix
  * Use default level over set #9

  @odyssey4me contributed:
  * Array `snmp['snmpd']['mibs']` to specify list of MIBs to load #12
  * Arrays `snmp['process_monitory']['proc']` and `snmp['process_monitoring']['procfix']
    for monitoring process table and corrective procfix commands. #13
  * Attributes `snmp['disman_events']` for SNMP Distributed Management #14

  @sfiggins contributed:
  * Attributes `snmp['include_all_disks']` (bool),
    `snmp['all_disk_min']` (String for minimum kilobytes/percent free),
    and `snmp['disks']` (Hash of mount points and/or minimum thresholds). #17

  * Corrected rubocop warnings.
  * Add chefspec/Remove minitest
  * Add basic BATS tests
  * Add erubis check for templates

v1.0.0
------
  Add snmptrapd recipe, and add RHEL support.
  Template debian files.
  Add test-kitchen skeleton.
  Typos in platform_family case switches.
  Check for existence of dmi OHAI attribute, before checking in a condition.
  Correct minitest-spec file for Chef 11.x

v0.4.0
------
  Add necessary setup for the HP System Management Homepage to be able to poll via @tas50
  Add Suse support via @jackl0phty
  FC043: Prefer new notification syntax: ./recipes/default.rb:38

v0.3.1
------
Fix Gemfile for Travis.

Foodcritic warnings resolved

* FC007: Ensure recipe dependencies are reflected in cookbook metadata

v0.3.0
------
Merge pull request #1 from gustavowt/master

* Ubuntu package name fixes

Add [test-kitchen](https://github.com/opscode/test-kitchen) scaffolding

* Simple snmp::default test

Add TravisCI attribute sanity tests, and foodcritic linting.

Foodcritic warnings resolved

* FC001: Use strings in preference to symbols to access node attributes
* FC024: Consider adding platform equivalents

v0.2.0
------
Refactor SNMP cookbook

v0.1.0
------
* Add Getopt::Declare dependency
* Update metadata
* Update documentation

v0.0.1
------
Add snmp cookbook.


test/integration/default/inspec/default_test.rb
# # encoding: utf-8

pkg = if os.debian?
        'snmp snmpd'
      elsif os.redhat?
        'net-snmp net-snmp-utils'
      elsif os.suse?
        'net-snmp'
      elsif os.linux?
        'net-snmpd'
      elsif os.bsd?
        'net-snmp'
      else
        'net-snmp net-snmp-utils'
      end

file = if os.debian?
         '/etc/snmp/snmpd.conf'
       elsif os.redhat?
         '/etc/snmp/snmpd.conf'
       elsif os.suse?
         '/etc/snmp/snmpd.conf'
       elsif os.linux?
         '/etc/snmp/snmpd.conf'
       elsif os.bsd?
         ' /usr/local/share/snmp/snmpd.conf'
       end

describe package(pkg) do
  it { should be_installed }
end

describe service('snmpd') do
  it { should be_installed }
  it { should be_running }
end

if os.redhat?
    describe service('snmptrapd') do
      it { should be_installed }
      it { should be_running }
    end
end

describe file(file) do
  it { should exist }
  it { should be_owned_by 'root' }
  its('type') { should eq :file }
  it { should be_file }
  it { should_not be_directory }
  its('mode') { should cmp '0600' }
end


recipes/snmptrapd.rb
#
# Cookbook Name:: snmp
# Recipe:: snmptrapd
#
# Copyright 2013, Eric G. Wolfe
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

node.default['snmp']['snmptrapd']['trapd_run'] = 'yes'

include_recipe 'snmp'

service node['snmp']['snmptrapd']['service'] do
  action [:enable, :start]
end

template '/etc/snmp/snmptrapd.conf' do
  mode 0644
  owner 'root'
  group 'root'
  notifies :restart, "service[#{node['snmp']['snmptrapd']['service']}]"
end


attributes/default.rb
#
# Cookbook Name:: snmp
# Attributes:: default
#
# Copyright 2010, Eric G. Wolfe
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# family specific config
case node['platform_family']
when 'rhel'
  default['snmp']['packages'] = %w(net-snmp net-snmp-utils)
  default['snmp']['snmpd']['snmpd_opts'] = '-LS0-5d -Lf /dev/null -p /var/run/snmpd.pid'
when 'debian'
  default['snmp']['packages'] = %w(snmp snmpd)
  default['snmp']['snmpd']['snmpd_run'] = 'yes'
  default['snmp']['snmpd']['snmpd_opts'] = '-Lsd -Lf /dev/null -u snmp -g snmp -I -smux -p /var/run/snmpd.pid'
  # default['snmp']['snmptrapd']['trapd_run'] = 'no'
  default['snmp']['snmptrapd']['trapd_opts'] = '-Lsd -p /var/run/snmptrapd.pid'
  default['snmp']['snmpd']['snmpd_compat'] = 'yes'
when 'suse'
  default['snmp']['packages'] = %w(net-snmp)
when 'fedora'
  default['snmp']['packages'] = %w(net-snmp net-snmp-utils)
else
  default['snmp']['packages'] = %w(net-snmp net-snmp-utils)
end

# Same on supported platforms:
# redhat, centos, fedora, scientific, debian, ubuntu
default['snmp']['service'] = 'snmpd'

# Global snmp config
default['snmp']['mibdirs'] = '/usr/share/snmp/mibs'
default['snmp']['mibs'] = nil

# snmpd - basis
default['snmp']['snmpd']['conffile'] = '/etc/snmp/snmpd.conf'
default['snmp']['snmpd']['conf_owner'] = 'root'
default['snmp']['snmpd']['conf_group'] = 'root'
default['snmp']['snmpd']['conf_mode'] = '0600'

# snmpd - inside config
default['snmp']['snmpd']['agentAddress'] = 'udp:161'
default['snmp']['snmpd']['agentAddress6'] = 'udp6:161'
default['snmp']['snmpd']['community'] = 'public'
default['snmp']['snmpd']['sec_name'] = { 'notConfigUser' => %w(default) }
default['snmp']['snmpd']['sec_name6'] = { 'notConfigUser' => %w(default) }
default['snmp']['snmpd']['groups']['v1'] = { 'notConfigGroup' => %w(notConfigUser) }
default['snmp']['snmpd']['groups']['v2c'] = { 'notConfigGroup' => %w(notConfigUser) }
default['snmp']['snmpd']['full_systemview'] = false
default['snmp']['snmpd']['additional_oids'] = nil # Might be String or Array of Strings
default['snmp']['snmpd']['syslocationVirtual'] = 'Virtual Server'
default['snmp']['snmpd']['syslocationPhysical'] = 'Server Room'
default['snmp']['snmpd']['syscontact'] = 'Root <root@localhost>'
default['snmp']['snmpd']['sysname'] = node['fqdn']
default['snmp']['snmpd']['pass'] = {}
default['snmp']['snmpd']['pass_persist'] = {}
default['snmp']['snmpd']['process_monitoring']['proc'] = []
default['snmp']['snmpd']['process_monitoring']['procfix'] = []
default['snmp']['snmpd']['include_all_disks'] = false
default['snmp']['snmpd']['all_disk_min'] = 100    # 100K
default['snmp']['snmpd']['disks'] = []
default['snmp']['snmpd']['ignoredisks'] = []
default['snmp']['snmpd']['skipNFSInHostResources'] = nil
default['snmp']['snmpd']['load_average'] = []
default['snmp']['snmpd']['trapsinks'] = []
default['snmp']['snmpd']['dontLogTCPWrappersConnects'] = nil
default['snmp']['snmptrapd']['trapcommunity'] = 'public'
default['snmp']['snmptrapd']['trapd_run'] = 'no'
default['snmp']['snmpd']['extend_scripts'] = {}
default['snmp']['snmpd']['disman_events']['enable'] = false
default['snmp']['snmpd']['disman_events']['user'] = 'disman_events'
default['snmp']['snmpd']['disman_events']['password'] = 'disman_password'
default['snmp']['snmpd']['disman_events']['linkUpDownNotifications'] = 'yes'
default['snmp']['snmpd']['disman_events']['defaultMonitors'] = 'yes'
default['snmp']['snmpd']['disman_events']['monitors'] = []
